src/object/cli_ec.c

```diff
@@ -33,6 +33,454 @@
+	for (i = 0; i < (oca)->u.ec.e_p; i++) {				       \
+		seg_nrs[(oca)->u.ec.e_k + i] += (seg_num);		       \
+	}
+
+static int
+obj_ec_riod_init(daos_iod_t *riod, uint32_t recx_nr, bool with_eprs)
+{
+	D_ALLOC_ARRAY(riod->iod_recxs, recx_nr);
+	if (riod->iod_recxs == NULL)
+		return -DER_NOMEM;
+	if (!with_eprs)
+		return 0;
+	D_ALLOC_ARRAY(riod->iod_eprs, recx_nr);
+	if (riod->iod_eprs == NULL)
+		return -DER_NOMEM;
+	return 0;
+}
+
+/** scan the iod to find the full_stripe recxs and some help info */
+int
+obj_ec_recx_scan(daos_iod_t *iod, d_sg_list_t *sgl,
+		 struct daos_oclass_attr *oca, struct obj_reasb_req *reasb_req,
+		 uint32_t iod_idx)
+{
+	uint8_t				*tgt_bitmap = reasb_req->tgt_bitmap;
+	struct obj_ec_recx_array	*ec_recx_array;
+	struct obj_ec_recx		*ec_recx = NULL;
+	daos_recx_t			*recx;
+	uint32_t			*tgt_recx_nrs, *tgt_seg_nrs;
+	uint32_t			 recx_nr, tgt_nr, seg_nr;
+	uint32_t			 oiod_flags = 0;
+	uint64_t			 stripe_rec_nr;
+	uint64_t			 start, end, rec_nr, rec_off;
+	bool				 full_stripe_only = true;
+	int				 i, j, idx, rc;
+
+	D_ASSERT(iod->iod_type == DAOS_IOD_ARRAY);
+	stripe_rec_nr = obj_ec_stripe_rec_nr(oca);
+	ec_recx_array = &reasb_req->orr_recxs[iod_idx];
+	tgt_recx_nrs = ec_recx_array->oer_tgt_recx_nrs;
+	tgt_seg_nrs = ec_recx_array->oer_tgt_seg_nrs;
+	ec_recx_array->oer_k = oca->u.ec.e_k;
+	ec_recx_array->oer_p = oca->u.ec.e_p;
+
+	for (i = 0, idx = 0, rec_off = 0; i < iod->iod_nr; i++) {
+		recx = &iod->iod_recxs[i];
+		start = roundup(recx->rx_idx, stripe_rec_nr);
+		end = rounddown(recx->rx_idx + recx->rx_nr, stripe_rec_nr);
+		if (start >= end) {
+			ec_partial_tgt_seg_recx(recx, stripe_rec_nr, oca,
+						tgt_recx_nrs, tgt_seg_nrs, j);
+			rec_off += recx->rx_nr;
+			full_stripe_only = false;
+			continue;
+		}
+
+		if (ec_recx_array->oer_recxs == NULL) {
+			rc = obj_ec_recxs_init(ec_recx_array, iod->iod_nr - i);
+			if (rc)
+				return rc;
+			ec_recx = ec_recx_array->oer_recxs;
+		}
+		ec_recx[idx].oer_idx = i;
+		rec_nr = end - start;
+		ec_recx[idx].oer_stripe_nr = rec_nr / stripe_rec_nr;
+		ec_recx[idx].oer_byte_off = (rec_off + start - recx->rx_idx) *
+					    iod->iod_size;
+		ec_recx[idx].oer_recx.rx_idx = start;
+		ec_recx[idx].oer_recx.rx_nr = rec_nr;
+		ec_recx_array->oer_stripe_total += ec_recx[idx].oer_stripe_nr;
+		idx++;
+		rec_off += recx->rx_nr;
+		/* at least one recx on each tgt for full stripe */
+		ec_all_tgt_recx_nrs(oca, tgt_recx_nrs, j);
+		/* update seg_nrs on data cells */
+		ec_data_tgt_nrs(recx, tgt_recx_nrs, tgt_seg_nrs, oca, 0, 1);
+		/* update seg_nrs on parity cells for full stripe */
+		ec_parity_tgt_seg_nrs(oca, tgt_seg_nrs,
+				     ec_recx[idx].oer_stripe_nr, j);
+		/* partial update before or after full stripe need replica to
+		 * parity target.
+		 */
+		if (recx->rx_idx < start)
+			ec_replica_pairty_seg_recx(oca, tgt_recx_nrs,
+						   tgt_seg_nrs, j);
+		if (recx->rx_idx + recx->rx_nr > end)
+			ec_replica_pairty_seg_recx(oca, tgt_recx_nrs,
+						   tgt_seg_nrs, j);
```

Comment:
the reason is we need a compact recx array for all targets, so if one IO only involved partial targets can get a compact array. For example an IO to target 0, 1, 6, 7, then in the new iod's iod_recxs array it will have an array to include those targets' recxs and need not add zero bytes for 2~5. So I scan it firstly to get some basic info like which set of target involved and how many recxs needed for each targets. Same for sgl.
we may discuss offline, thanks